db/snapshot-ctl.cc

```diff
@@ -136,18 +136,20 @@ future<int64_t> snapshot_ctl::true_snapshots_size() {
     }));
 }
 
-future<tasks::task_id> snapshot_ctl::start_backup(sstring endpoint, sstring bucket, sstring prefix, sstring keyspace, sstring snapshot_name) {
+future<tasks::task_id> snapshot_ctl::start_backup(sstring endpoint, sstring bucket, sstring prefix, sstring keyspace, sstring table, sstring snapshot_name) {
     if (this_shard_id() != 0) {
         co_return co_await container().invoke_on(0, [&](auto& local) {
-            return local.start_backup(endpoint, bucket, prefix, keyspace, snapshot_name);
+            return local.start_backup(endpoint, bucket, prefix, keyspace, table, snapshot_name);
         });
     }
 
     co_await coroutine::switch_to(_config.backup_sched_group);
     auto cln = _storage_manager.get_endpoint_client(endpoint);
     snap_log.info("Backup sstables from {}({}) to {}", keyspace, snapshot_name, endpoint);
+    auto table_id = _db.local().find_uuid(keyspace, table);
+    auto table_dir = replica::format_table_directory_name(table, table_id);;
```

Comment:
Please, avoid using this method, it's going to be removed.
Instead, you should find a table object itself, then grab storage options for it, and get the cf directory from one, like in #20675:

```patch
--- a/replica/table.cc
+++ b/replica/table.cc
@@ -2595,7 +2595,15 @@ future<> table::write_schema_as_cql(database& db, sstring dir) const {
 
 // Runs the orchestration code on an arbitrary shard to balance the load.
 future<> table::snapshot_on_all_shards(sharded<database>& sharded_db, const global_table_ptr& table_shards, sstring name) {
-    auto jsondir = table_shards->_config.datadir + "/snapshots/" + name;
+    auto* so = std::get_if<storage_options::local>(&table_shards->get_storage_options().value);
+    if (so == nullptr) {
+        throw std::runtime_error("Snapshotting non-local tables is not implemented");
+    }
+    if (so->dir.empty()) { // virtual tables don't have initialized local storage
+        co_return;
+    }
+
+    auto jsondir = (so->dir / sstables::snapshots_dir / name).native();
     auto orchestrator = std::hash<sstring>()(jsondir) % smp::count;
 
     co_await smp::submit_to(orchestrator, [&] () -> future<> {
```

Reply:
> @xemul you meant that format_table_directory_name() is going to be removed? 

Yes. replica::format_table_directory_name() is going to be removed. It will find its place in sstables namespace (and one of related .cc file)

> but in https://github.com/scylladb/scylladb/pull/20675, what is removed is datadir. and format_table_directory_name() is not using this field. 

Yes

> or, are you referencing _snap_ctl._db.local().get_config().data_file_directories() ?

No, try not to use it either :) If you want to know where a `table` keeps its files, you should use table.get_storage_options() and, if it's storage_options::local, use its `.dir` field

Reply:
@xemul if i would need to replace `format_table_directory_name()` for preparing for the change you are talking about i would need to replicate it -- it's but a 3 liner. it's fine, but once your change dropping this function, i will need to replace the 3-liner with whatever it's replaced with. do you think it's a good plan? i doubt it.

to drop the usage of `data_file_directories()` was not in the scope of this change, but i will include it anyway.